src/common/src/hash/key.rs

```diff
@@ -76,10 +76,15 @@ pub trait HashKeyDeserializer {
     fn deserialize<'a, D: HashKeySerDe<'a>>(&'a mut self) -> ArrayResult<Option<D>>;
 }
 
+/// Trait for value types that can be serialized to or deserialized from hash keys.
+///
+/// Note that this trait is more like a marker suggesting that types that implement it can be
+/// encoded into the hash key. The actual encoding/decoding method is not limited to
+/// [`HashKeySerDe`]'s fixed-size implementation.
 pub trait HashKeySerDe<'a>: ScalarRef<'a> {
     type S: AsRef<[u8]>;
-    fn serialize(self) -> Self::S;
-    fn deserialize<R: Read>(source: &mut R) -> Self;
+    fn fixed_size_serialize(self) -> Self::S;
```

Comment:
I don't think we should rename this. After you serialized no fixed size to `Vec<u8>`, what shoud be changed is the signature 
```
fn serialize(self, buffer: &mut W: Writer)
```
And after that, we can implement `serialized`/`deserialize` for all value types.

Reply:
This rename is a temporary one, aimed to reflect the fixed-sizedness of the implementation. It should definitely be refactored later.

I think this `Writer` is an opportunity to unify the interface for key/value/fixed size encoding, but I guess it will be a big refactor.